icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -483,37 +545,50 @@ class Parser {
                     status = kUnitIdentifierSyntaxError;
                     return;
             }
-        }
 
-        // We ran out of tokens before finding a complete single unit.
-        status = kUnitIdentifierSyntaxError;
+            previ = fIndex;
+            token = nextToken(status);
+            if (U_FAILURE(status)) {
+                // We have run out of (valid) tokens before finding a complete
+                // single unit.
+                return;
+            }
+        }
     }
 
+    /// @param result is modified, not overridden. Caller must pass in a
+    /// default-constructed (empty) MeasureUnitImpl instance.
     void parseImpl(MeasureUnitImpl& result, UErrorCode& status) {
         if (U_FAILURE(status)) {
             return;
         }
+        if (fSource.empty()) {
+            // The dimenionless unit: nothing to parse. leave result as is.
+            return;
+        }
         int32_t unitNum = 0;
+        if (!hasNext()) {
+            // We don't support the empty string.
```

Comment:
Not relevant any more.  You could `U_ASSERT(hasNext())` or just remove this if statement.